File: Src/java/engine-fhir/src/test/java/org/hl7/fhirpath/TranslatorHelper.java
package org.hl7.fhirpath;

import java.util.ArrayList;
import java.util.List;
import org.cqframework.cql.cql2elm.*;
import org.cqframework.cql.cql2elm.CqlCompilerOptions.Options;
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider;
import org.cqframework.cql.cql2elm.tracking.TrackBack;
import org.fhir.ucum.UcumException;
import org.hl7.elm.r1.Library;
import org.opencds.cqf.cql.engine.execution.CqlEngine;
import org.opencds.cqf.cql.engine.execution.Environment;
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;

public class TranslatorHelper {

    private TranslatorHelper() {
        // intentionally empty
    }

    private static ModelManager modelManager = new ModelManager();
    private static LibraryManager libraryManager;

    private static LibraryManager getLibraryManager() {
        var options = CqlCompilerOptions.defaultOptions();
        options.getOptions().remove(Options.DisableListDemotion);
        options.getOptions().remove(Options.DisableListPromotion);
        options.getOptions().add(Options.EnableDateRangeOptimization);
        return getLibraryManager(options);
    }

    private static LibraryManager getLibraryManager(CqlCompilerOptions cqlCompilerOptions) {
        libraryManager = new LibraryManager(modelManager, cqlCompilerOptions);
        libraryManager.getLibrarySourceLoader().clearProviders();
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
        libraryManager.getLibrarySourceLoader().registerProvider(new FhirLibrarySourceProvider());
        return libraryManager;
    }

    public static CqlEngine getEngine(String cql) {
        return getEngine(getEnvironment(cql));
    }

    public static Environment getEnvironment(String cql) {
        var env = getEnvironment((TerminologyProvider) null);
        env.getLibraryManager().getCompiledLibraries().clear();
        env.getLibraryManager()
                .getLibrarySourceLoader()
                .registerProvider(new StringLibrarySourceProvider(List.of(cql)));
        return env;
    }

    public static Environment getEnvironment() {
        return getEnvironment((TerminologyProvider) null);
    }

    public static Environment getEnvironment(TerminologyProvider terminologyProvider) {
        return new Environment(getLibraryManager(), null, terminologyProvider);
    }

    public static CqlEngine getEngine(Environment environment) {
        return new CqlEngine(environment);
    }

    public static org.hl7.elm.r1.VersionedIdentifier toElmIdentifier(String name) {
        return new org.hl7.elm.r1.VersionedIdentifier().withId(name);
    }

    public static org.hl7.elm.r1.VersionedIdentifier toElmIdentifier(String name, String version) {
        return new org.hl7.elm.r1.VersionedIdentifier().withId(name).withVersion(version);
    }

    public static Library translate(String cql, LibraryManager libraryManager) throws UcumException {
        CqlCompiler compiler = new CqlCompiler(libraryManager);
        Library lib = compiler.run(cql);

        libraryManager.getCompiledLibraries().put(lib.getIdentifier(), compiler.getCompiledLibrary());

        if (!compiler.getErrors().isEmpty()) {
            ArrayList<String> errors = new ArrayList<>();
            for (CqlCompilerException error : compiler.getErrors()) {
                TrackBack tb = error.getLocator();
                String lines = tb == null
                        ? "[n/a]"
                        : String.format(
                                "[%d:%d, %d:%d]",
                                tb.getStartLine(), tb.getStartChar(), tb.getEndLine(), tb.getEndChar());
                errors.add(lines + error.getMessage());
            }
            throw new IllegalArgumentException(errors.toString());
        }

        return lib;
    }
}

File: Src/java/engine-fhir/src/test/kotlin/org/hl7/fhirpath/TranslatorHelper.kt
package org.hl7.fhirpath

import org.cqframework.cql.cql2elm.CqlCompiler
import org.cqframework.cql.cql2elm.CqlCompilerException.ErrorSeverity
import org.cqframework.cql.cql2elm.CqlCompilerOptions
import org.cqframework.cql.cql2elm.CqlCompilerOptions.Companion.defaultOptions
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.cqframework.cql.cql2elm.StringLibrarySourceProvider
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider
import org.fhir.ucum.UcumException
import org.hl7.elm.r1.Library
import org.hl7.elm.r1.VersionedIdentifier
import org.opencds.cqf.cql.engine.execution.CqlEngine
import org.opencds.cqf.cql.engine.execution.Environment
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider

object TranslatorHelper {
    private val modelManager = ModelManager()
    private var libraryManager: LibraryManager? = null

    private fun getLibraryManager(): LibraryManager {
        val options = defaultOptions()
        options.options.remove(CqlCompilerOptions.Options.DisableListDemotion)
        options.options.remove(CqlCompilerOptions.Options.DisableListPromotion)
        options.options.add(CqlCompilerOptions.Options.EnableDateRangeOptimization)
        return getLibraryManager(options)
    }

    private fun getLibraryManager(cqlCompilerOptions: CqlCompilerOptions): LibraryManager {
        libraryManager = LibraryManager(modelManager, cqlCompilerOptions)
        libraryManager!!.librarySourceLoader.clearProviders()
        libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        libraryManager!!.librarySourceLoader.registerProvider(FhirLibrarySourceProvider())
        return libraryManager!!
    }

    @JvmStatic
    fun getEngine(cql: String): CqlEngine {
        return getEngine(getEnvironment(cql))
    }

    fun getEnvironment(cql: String): Environment {
        val env = getEnvironment(null as TerminologyProvider?)
        env.libraryManager!!.compiledLibraries.clear()
        env.libraryManager!!
            .librarySourceLoader
            .registerProvider(StringLibrarySourceProvider(listOf(cql)))
        return env
    }

    val environment: Environment
        get() = getEnvironment(null as TerminologyProvider?)

    fun getEnvironment(terminologyProvider: TerminologyProvider?): Environment {
        return Environment(getLibraryManager(), null, terminologyProvider)
    }

    fun getEngine(environment: Environment?): CqlEngine {
        return CqlEngine(environment!!)
    }

    @JvmStatic
    fun toElmIdentifier(name: String?): VersionedIdentifier {
        return VersionedIdentifier().withId(name)
    }

    @JvmStatic
    fun toElmIdentifier(name: String?, version: String?): VersionedIdentifier {
        return VersionedIdentifier().withId(name).withVersion(version)
    }

    @Throws(UcumException::class)
    fun translate(cql: String?, libraryManager: LibraryManager): Library {
        val compiler = CqlCompiler(libraryManager = libraryManager)
        val lib = compiler.run(cql!!)

        libraryManager.compiledLibraries[lib.identifier!!] = compiler.compiledLibrary!!

        val errors = compiler.exceptions.filter { it.severity == ErrorSeverity.Error }
        if (errors.isNotEmpty()) {
            val messages = mutableListOf<String>()
            for (error in compiler.exceptions) {
                val tb = error.locator
                val lines =
                    if (tb == null) "[n/a]"
                    else
                        String.format(
                            "[%d:%d, %d:%d]",
                            tb.startLine,
                            tb.startChar,
                            tb.endLine,
                            tb.endChar,
                        )
                messages.add(lines + error.message)
            }
            throw IllegalArgumentException(messages.toString())
        }

        return lib
    }
}
